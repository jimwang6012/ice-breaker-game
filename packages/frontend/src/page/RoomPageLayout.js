import React, { useContext, useEffect, useRef, useState } from "react";

import { Outlet, useNavigate } from "react-router-dom";
import { Text, ScrollArea, Group } from "@mantine/core";
import { useInputState, useListState } from "@mantine/hooks";
import { createStyles, Avatar } from "@mantine/core";
import { AppContext } from "../AppContextProvider";
import socket from "../Socket";
import { Modal } from "../component/Modal";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCheck } from "@fortawesome/free-solid-svg-icons";
import { RiInformationFill, RiVipCrownFill } from "react-icons/ri";

export default function RoomPageLayout() {
  const navigate = useNavigate();
  const { classes } = useStyles();
  const { roomId, toggle, stop } = useContext(AppContext);
  const [show, setShow] = useState(false);
  const hostLeftInfo = "Host has left the room! The room is closed.";
  const returnToHomePrompt = "Return to home";

  // On browser back button click
  const onBackButtonEvent = (e) => {
    e.preventDefault();
    socket.disconnect();
    stop();
    navigate("/home");
  };

  useEffect(() => {
    // When host leave, inform the other players that the room is closed.
    const openModal = () => {
      setShow(true);
    };
    socket.on("room-closed", openModal);
    toggle();
    // Navigate to home page if context is lost, eg. on refresh.
    if (!roomId) {
      navigate("/home");
    }

    window.history.pushState(null, null, window.location.pathname);
    window.addEventListener("popstate", onBackButtonEvent);
    return () => {
      // Clean up
      socket.off("room-closed", openModal);
      window.removeEventListener("popstate", onBackButtonEvent);
    };
  }, []);

  if (!roomId) {
    return null;
  }

  return (
    <div className={classes.roomPage}>
      {/* This is the section of the room layout that shows the game page and idle page. */}
      <div className={classes.leftSection}>
        <Outlet />
      </div>
      {/* This is the section of the room layout that shows the player list and message list */}
      <div className={classes.rightSection}>
        <PlayerList />
        <MessageList />
      </div>
      {/* Modal component when room closed */}
      <Modal
        show={show}
        pageJump={() => {
          navigate("/home");
        }}
        mainPrompt={hostLeftInfo}
        buttonPrompt={returnToHomePrompt}
        title={
          <div className="pl-[43%] pb-5 text-5xl font-black text-white center text-ice-6">
            <RiInformationFill />
          </div>
        }
      />
    </div>
  );
}

//player list componenet
function PlayerList() {
  const { classes } = useStyles();
  const { handlers, hostId, name, players, colors } = useContext(AppContext);
  const updateGame = (data) => {
    handlers.setState(data.players);
  };
  useEffect(() => {
    socket.on("game-update", updateGame);
    return () => {
      socket.off("game-update", updateGame);
    };
  }, []);

  return (
    <div className={classes.playerList}>
      <Text size="lg" style={{ paddingBottom: 8 }}>
        Players ({players.length})
      </Text>
      <ScrollArea
        scrollbarSize={8}
        offsetScrollbars
        classNames={{
          root: classes.scrollArea,
          scrollbar: classes.scrollbar,
          thumb: classes.thumb,
        }}
      >
        {players.map((p, index) => (
          <PlayerItem
            key={index}
            color={colors[p.colorId]}
            selfName={name}
            playerName={p.name}
            isHost={p.id === hostId}
            isBreaker={p.isBreaker}
            isReady={p.isReady}
          />
        ))}
      </ScrollArea>
    </div>
  );
}

//player list item
function PlayerItem({
  color,
  selfName,
  playerName,
  isHost,
  isBreaker,
  isReady,
}) {
  const { classes } = useStyles();
  return (
    <Group spacing={"xs"} className={classes.playerItem}>
      <FontAwesomeIcon
        className={`text-green-500 ${isReady ? "visible" : "invisible"}`}
        icon={faCheck}
      />
      <Avatar
        src={
          isBreaker
            ? process.env.PUBLIC_URL + "/avatars/seal_down.png"
            : process.env.PUBLIC_URL + "/avatars/p_down_black.png"
        }
        size="md"
        radius="md"
        styles={{
          image: {
            backgroundColor: color ? color : "white",
          },
        }}
      />
      <Text size="xl" weight={500}>
        {playerName === selfName ? playerName + " (Me)" : playerName}
      </Text>
      <div className={isHost ? "visible" : "invisible"}>
        <RiVipCrownFill color="#f4ac3f" />
      </div>
    </Group>
  );
}

//message list componenet
function MessageList() {
  const { classes } = useStyles();
  const [messageList, handlers] = useListState([]);
  const { roomId, name, setIsTyping, players, colors } = useContext(AppContext);

  const receiveMessage = (message) => {
    handlers.append(message);
    scrollToBottom();
  };

  useEffect(() => {
    socket.on("message-to-chat", receiveMessage);
    return () => {
      socket.off("message-to-chat", receiveMessage);
    };
  }, [socket]);

  const sendMessage = (message) => {
    socket.emit("send-message", {
      roomID: roomId,
      chatMessage: name + ": " + message,
    });
  };

  const viewport = useRef();

  // scroll to bottom of the message list
  const scrollToBottom = () =>
    viewport.current.scrollTo({
      top: viewport.current.scrollHeight,
      behavior: "smooth",
    });

  return (
    <div className={classes.messageList}>
      <ScrollArea
        id="scroll"
        scrollbarSize={8}
        offsetScrollbars
        classNames={{
          root: classes.scrollArea,
          scrollbar: classes.scrollbar,
          thumb: classes.thumb,
        }}
        viewportRef={viewport}
      >
        {messageList.map((m, index) => {
          if (m.type === "chat") {
            const msgArray = m.value.split(":");
            const msgName = msgArray[0];
            const playerColor = players.filter(
              (player) => player.name === msgName
            )[0].colorId;
            return (
              <MessageItem
                key={index}
                value={m.value}
                colors={colors}
                colorId={playerColor}
              />
            );
          } else if (m.type === "system") {
            return <NotiItem key={index} value={m.value} />;
          }
        })}
      </ScrollArea>
      {/* Text input box */}
      <div className={classes.InputArea}>
        <input
          className={classes.InputField}
          onFocus={() => {
            setIsTyping(true);
          }}
          onBlur={() => {
            setIsTyping(false);
          }}
          onKeyPress={(e) => {
            if (e.key === "Enter") {
              sendMessage(e.currentTarget.value);
              e.currentTarget.value = "";
            }
          }}
        ></input>
      </div>
    </div>
  );
}

//System messages
function NotiItem({ value }) {
  const { classes } = useStyles();
  return (
    <Text size="xl" weight={500} className={classes.NotiItem}>
      {value}
    </Text>
  );
}

function MessageItem({ value, colors, colorId }) {
  const { classes } = useStyles();
  let msgColor;
  if (colorId !== null) {
    msgColor = colors[colorId];
  } else {
    msgColor = "#ffffff";
  }
  return (
    <div className="flex flex-row items-start pl-3">
      <div
        className={`w-0.5 h-4 mt-1 rounded-md`}
        style={{ backgroundColor: msgColor }}
      />
      <Text size="md" className={classes.MessageItem}>
        {value}
      </Text>
    </div>
  );
}

// Create styles using Mantine theme color
const useStyles = createStyles((theme) => ({
  playerList: {
    backgroundColor: theme.colors.ice[6],
    height: "45vh",
    color: "white",
    paddingInline: 10,
  },
  InputArea: {
    backgroundColor: theme.colors.ice[7],
    height: "6vh",
  },
  messageList: {
    "& #scroll": {
      height: "100%",
    },
    backgroundColor: theme.colors.ice[7],
    height: "49vh",
    color: "white",
  },
  roomPage: {
    display: "flex",
  },
  leftSection: {
    width: "68.75%",
    height: "100vh",
  },
  rightSection: {
    width: "31.25%",
    height: "100vh",
  },

  scrollArea: {
    height: "91%",
  },

  scrollbar: {
    backgroundColor: "transparent",
    "&:hover": {
      backgroundColor: "transparent",
    },
  },
  thumb: {
    backgroundColor: theme.colors.ice[2],
    "&:hover": {
      backgroundColor: theme.colors.ice[2],
    },
  },
  playerItem: {
    marginBottom: 8,
  },
  NotiItem: {
    backgroundColor: theme.colors.ice[5],
    paddingInline: 10,
    marginTop: 5,
    marginBottom: 5,
  },
  MessageItem: {
    paddingInline: 3,
    wordBreak: "break-all",
    flex: 1,
    flexWrap: "wrap",
    color: theme.colors.white,
  },
  InputField: {
    backgroundColor: theme.colors.ice[6],
    margin: 8,
    borderRadius: 6,
    paddingInline: 8,
    color: "white",
    width: "96%",
    onKeyDown: "if(event.keyCode==13)",
    type: "text",
    height: "70%",
  },
}));
