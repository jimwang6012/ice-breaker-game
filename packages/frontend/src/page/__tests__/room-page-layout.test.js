import { BrowserRouter } from "react-router-dom";
import RoomPageLayout from "../RoomPageLayout";
import { MantineProvider } from "@mantine/core";

import { render } from "@testing-library/react";

it("renders correctly", () => {
  class ResizeObserver {
    observe() {}
    unobserve() {}
  }

  // @ts-ignore
  window.ResizeObserver = ResizeObserver;
  const tree = render(
    <BrowserRouter>
      <MantineProvider
        emotionOptions={{ key: "mantine", prepend: false }}
        // @ts-ignore
        withCSSVariables
        theme={{
          colors: {
            ice: [
              "#DDEEF8",
              "#C5E0F5",
              "#9EBFE0",
              "#5C83A0",
              "#04AED9",
              "#0478A1",
              "#003853",
              "#041F32",
              "#768EAA",
              "#768EAA",
            ],
          },
        }}
        styles={{
          Button: (theme) => ({
            filled: {
              // subscribe to component params
              color: "white",
              background: theme.colors.ice[6],
              "&:hover": {
                backgroundColor: theme.colors.ice[6],
              },
            },
          }),
        }}
      >
        <RoomPageLayout />
      </MantineProvider>
    </BrowserRouter>
  );
  expect(tree).toMatchSnapshot();
});
