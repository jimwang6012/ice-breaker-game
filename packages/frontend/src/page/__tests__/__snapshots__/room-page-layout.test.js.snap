// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mantine-17do188"
      >
        <div
          class="mantine-18mbv9t"
        />
        <div
          class="mantine-a5dh29"
        >
          <div
            class="mantine-14p7xee"
          >
            <div
              class="mantine-Text-root mantine-1k8fja7"
              style="padding-bottom: 5px;"
            >
              Players (
              0
              )
            </div>
            <div
              class="mantine-ScrollArea-root mantine-1fnbhs0"
              dir="ltr"
              style="position: relative; --radix-scroll-area-corner-width: 0px; --radix-scroll-area-corner-height: 0px;"
            >
              <style>
                [data-radix-scroll-area-viewport]{scrollbar-width:none;-ms-overflow-style:none;-webkit-overflow-scrolling:touch;}[data-radix-scroll-area-viewport]::-webkit-scrollbar{display:none}
              </style>
              <div
                class="mantine-8dh50l mantine-ScrollArea-viewport"
                data-radix-scroll-area-viewport=""
                style="overflow-x: scroll; overflow-y: scroll;"
              >
                <div
                  style="min-width: 100%; display: table;"
                />
              </div>
            </div>
          </div>
          <div
            class="mantine-1bid5c6"
          >
             
            <div
              class="mantine-ScrollArea-root mantine-1fnbhs0"
              dir="ltr"
              id="scroll"
              style="position: relative; --radix-scroll-area-corner-width: 0px; --radix-scroll-area-corner-height: 0px;"
            >
              <style>
                [data-radix-scroll-area-viewport]{scrollbar-width:none;-ms-overflow-style:none;-webkit-overflow-scrolling:touch;}[data-radix-scroll-area-viewport]::-webkit-scrollbar{display:none}
              </style>
              <div
                class="mantine-8dh50l mantine-ScrollArea-viewport"
                data-radix-scroll-area-viewport=""
                style="overflow-x: scroll; overflow-y: scroll;"
              >
                <div
                  style="min-width: 100%; display: table;"
                />
              </div>
            </div>
            <div
              class="mantine-1svwv29"
            >
              <input
                class="mantine-1gnof2j"
              />
            </div>
          </div>
        </div>
        <div
          class="modal display-none"
        >
          <section
            class="modal-main"
          >
            <div
              class=""
            >
              <div
                class="p-10 rounded-lg shadow dark:bg-ice-8"
              >
                <div
                  class="text-center "
                >
                  <div
                    class="pl-[43%] pb-5 text-5xl font-black text-white center text-ice-6"
                  >
                    <svg
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <g>
                        <path
                          d="M0 0h24v24H0z"
                          fill="none"
                        />
                        <path
                          d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm-1-11v6h2v-6h-2zm0-4v2h2V7h-2z"
                        />
                      </g>
                    </svg>
                  </div>
                  <h3
                    class="mb-5 text-xl font-bold text-ice-7"
                  >
                    Host has left the room! The room is closed.
                  </h3>
                  <button
                    class="px-5 py-3 mt-5 font-semibold text-white rounded-lg text-l bg-ice-6 hover:bg-ice-5"
                  >
                    Return to home
                  </button>
                </div>
              </div>
            </div>
          </section>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mantine-17do188"
    >
      <div
        class="mantine-18mbv9t"
      />
      <div
        class="mantine-a5dh29"
      >
        <div
          class="mantine-14p7xee"
        >
          <div
            class="mantine-Text-root mantine-1k8fja7"
            style="padding-bottom: 5px;"
          >
            Players (
            0
            )
          </div>
          <div
            class="mantine-ScrollArea-root mantine-1fnbhs0"
            dir="ltr"
            style="position: relative; --radix-scroll-area-corner-width: 0px; --radix-scroll-area-corner-height: 0px;"
          >
            <style>
              [data-radix-scroll-area-viewport]{scrollbar-width:none;-ms-overflow-style:none;-webkit-overflow-scrolling:touch;}[data-radix-scroll-area-viewport]::-webkit-scrollbar{display:none}
            </style>
            <div
              class="mantine-8dh50l mantine-ScrollArea-viewport"
              data-radix-scroll-area-viewport=""
              style="overflow-x: scroll; overflow-y: scroll;"
            >
              <div
                style="min-width: 100%; display: table;"
              />
            </div>
          </div>
        </div>
        <div
          class="mantine-1bid5c6"
        >
           
          <div
            class="mantine-ScrollArea-root mantine-1fnbhs0"
            dir="ltr"
            id="scroll"
            style="position: relative; --radix-scroll-area-corner-width: 0px; --radix-scroll-area-corner-height: 0px;"
          >
            <style>
              [data-radix-scroll-area-viewport]{scrollbar-width:none;-ms-overflow-style:none;-webkit-overflow-scrolling:touch;}[data-radix-scroll-area-viewport]::-webkit-scrollbar{display:none}
            </style>
            <div
              class="mantine-8dh50l mantine-ScrollArea-viewport"
              data-radix-scroll-area-viewport=""
              style="overflow-x: scroll; overflow-y: scroll;"
            >
              <div
                style="min-width: 100%; display: table;"
              />
            </div>
          </div>
          <div
            class="mantine-1svwv29"
          >
            <input
              class="mantine-1gnof2j"
            />
          </div>
        </div>
      </div>
      <div
        class="modal display-none"
      >
        <section
          class="modal-main"
        >
          <div
            class=""
          >
            <div
              class="p-10 rounded-lg shadow dark:bg-ice-8"
            >
              <div
                class="text-center "
              >
                <div
                  class="pl-[43%] pb-5 text-5xl font-black text-white center text-ice-6"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g>
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                      <path
                        d="M12 22C6.477 22 2 17.523 2 12S6.477 2 12 2s10 4.477 10 10-4.477 10-10 10zm-1-11v6h2v-6h-2zm0-4v2h2V7h-2z"
                      />
                    </g>
                  </svg>
                </div>
                <h3
                  class="mb-5 text-xl font-bold text-ice-7"
                >
                  Host has left the room! The room is closed.
                </h3>
                <button
                  class="px-5 py-3 mt-5 font-semibold text-white rounded-lg text-l bg-ice-6 hover:bg-ice-5"
                >
                  Return to home
                </button>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
